### DISCUSSION TITLE
Dots on 2D map after deactivating the Tour Geo Filter

### DISCUSSION THREAD

**Discussion by FJBDev** [mytourbook/mytourbook#1308] [General]
Not sure if it's a bug or a feature but I have noticed that after I deactivate a Tour Geo Filter and select a tour in the TourBook view, the tour track is still displayed with dots.

Am i missing something ? is it normal ? If I restart MT, then the track is correctly displayed.

![image](https://github.com/mytourbook/mytourbook/assets/25235665/c364140a-ff3b-45e5-aaed-04fb08f4d7f8)

**Comment by wolfgang-ch** [comment]
I saw this bug also already but cannot reproduce it

**Reply by FJBDev** [reply]
I can reproduce it every time with the scenario I depicted above.
Should I create an official Bug for it ?

**Reply by wolfgang-ch** [reply]
> Should I create an official Bug for it ?

Yes

Do the bug also occur when

* just one perspective is open and
* all other views are closed except the tour book and 2D map view

**Reply by wolfgang-ch** [reply]
There is just one location where the fast painting is enabled/disabled https://github.com/mytourbook/mytourbook/blob/54b166d668a8a4e2e9d27ddc7e3a605cfba92d01/bundles/net.tourbook/src/net/tourbook/application/ActionTourGeoFilter.java#L74

**Comment by FJBDev** [comment]
> Do the bug also occur when
> 
>     * just one perspective is open and
> 
>     * all other views are closed except the tour book and 2D map view

yes, see below where even importing a GPX file (without saving it) shows the dots in the 2D map
![image](https://github.com/mytourbook/mytourbook/assets/25235665/09d9ce11-86d5-4398-8047-09114c6b3913)

**Reply by wolfgang-ch** [reply]
Is the black color OK for the track or is there another color when the track is displayed correctly?

**Reply by FJBDev** [reply]
Yes, the black color is OK because the track is mainly black when displayed correctly

**Comment by FJBDev** [comment]
> > Should I create an official Bug for it ?
> 
> Yes
> 

Bug created [here](https://github.com/mytourbook/mytourbook/issues/1310)

**Comment by wolfgang-ch** [comment]
> > Do the bug also occur when
> > ```
> > * just one perspective is open and
> > 
> > * all other views are closed except the tour book and 2D map view
> > ```
> 
> yes, see below where even importing a GPX file (without saving it) shows the dots in the 2D map ![image](https://private-user-images.githubusercontent.com/25235665/310579876-09d9ce11-86d5-4398-8047-09114c6b3913.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3OTMyMjgsIm5iZiI6MTcwOTc5MjkyOCwicGF0aCI6Ii8yNTIzNTY2NS8zMTA1Nzk4NzYtMDlkOWNlMTEtODZkNS00Mzk4LTgwNDctMDkxMTRjNmIzOTEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA3VDA2Mjg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwOTcwZWQ4ZjI5NWY4ZmI3ODI4NWU5ZWQwNzZjOGYwZmIwYWQxYjFhM2VhMThlMDMwMmI2MTUyYTY3YjVhMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1NCqVeG-L6FwNFycaIlS9-CacI2HdU5XrwiZoFNFwo8)

I see in your screenshot that other views are open, please ensure my above described setting and do this issue still occur?

Also restart MT with only these 2 opened views and verify this issue.

It is **VERY IMPORTANT** to know which views are open !!!

**Reply by FJBDev** [reply]
>Also restart MT with only these 2 opened views and verify this issue. 

Here is the same issue with your exact described settings (i.e.: `just one perspective is open and all other views are closed except the tour book and 2D map view`)

After I restarted, the issue disappeared

![image](https://github.com/mytourbook/mytourbook/assets/25235665/cd6f071c-c0ed-4c79-8bb3-c714e10eedb3)

**Comment by wolfgang-ch** [comment]
I tried to reproduce this issue in my 24.1 production, however it works as it should

https://github.com/mytourbook/mytourbook/assets/1283445/1ff6eaf2-d933-4c07-a300-d4ad31641824

**Reply by FJBDev** [reply]
This is what I see in my PROD MT 24.1 and the issue is persistent



https://github.com/mytourbook/mytourbook/assets/25235665/b75c9b3c-ad56-4a82-b0d0-689b16b5e80c

**Comment by wolfgang-ch** [comment]
Can you debug this code and check if the red marked code is run

![image](https://github.com/mytourbook/mytourbook/assets/1283445/f70f503a-bf74-4d1e-9b39-57a723a709cb)

**Reply by FJBDev** [reply]
>Can you debug this code and check if the red marked code is run

It's run when I activate or deactivate the Tour Geo filter

**Reply by wolfgang-ch** [reply]
I would now debug the code and would set a breakpoint at the field `_isFastMapPainting_Active` when it is modified. Then you could see when this field is set to `true`

![image](https://github.com/mytourbook/mytourbook/assets/1283445/b8a0858b-c793-48cd-a058-322f2588d4da)

**Reply by wolfgang-ch** [reply]
I hope you have not misunderstood my above comment "I would now debug the code...". I cannot debug the code in your IDE, you have to do it yourself because in cannot reproduce the bug.

A better wording would have been:

"You should now debug the code and set a breakpoint at the field `_isFastMapPainting_Active` when it is modified. Then you can see, when this field is set to `true`"

**Reply by FJBDev** [reply]
I think I understood correctly. But as of today, I don't know when I will have the opportunity to debug this.
I will update this thread once I have debugged

**Reply by FJBDev** [reply]
I took some time to do some investigation and while I still don't have a solid scenario to reproduce the issue every time, I believe I have found the culprit.

Using your debugging tip, I found that after deactivating the Geo Filter, [this](https://github.com/mytourbook/mytourbook/blob/67ab41215e5609a786612f5c13bd715f02018149/bundles/net.tourbook/src/net/tourbook/tour/filter/geo/TourGeoFilter_Loader.java#L137) line gets called making the issue appear.

My hunch is that it's a race issue where the task is [submitted ](https://github.com/mytourbook/mytourbook/blob/67ab41215e5609a786612f5c13bd715f02018149/bundles/net.tourbook/src/net/tourbook/tour/filter/geo/TourGeoFilter_Loader.java#L158)**before** the geo filter is deactivated but is actually **launched/executed** after the geo filter is deactivated and makes the dot appear even though the filter is already deactivated (and this is why it can't be reproduced every time)

**Reply by wolfgang-ch** [reply]
Can you test if this additional line could fix this issue

![image](https://github.com/mytourbook/mytourbook/assets/1283445/0cd66d12-fdb7-466b-9348-787fc426344b)

**Reply by wolfgang-ch** [reply]
Currently you should not do too much modifications in Map2.java because I'm currently implementing the map locations features with many modifications in this file.

**Comment by FJBDev** [comment]
> Can you test if this additional line could fix this issue
> 

After testing lots of time, i was not able to reproduce the issue with this code change. So maybe it's a valid fix ?
Maybe it's one of those issues that we will need to keep an eye on it but so far, I would say it's promising.
I will put it in my next PR

**Reply by wolfgang-ch** [reply]
When I investigated this issue then I discovered that this code is not very simple to understand and `Map2View.hideGeoGrid()` seems to be a fix, which did not cover all cases.

I'll add the extraline `_isFastMapPainting_Active = false;` into the code to prevent merge conflicts

**Comment by FJBDev** [comment]
> Currently you should not do too much modifications in Map2.java because I'm currently implementing the map locations features with many modifications in this file.

ok

**Reply by wolfgang-ch** [reply]
You should also not change related file, e.g.

![image](https://github.com/mytourbook/mytourbook/assets/1283445/632358e0-fc82-40a9-9856-ffaadb1dea2a)

otherwise there will be a lot of merge conflicts which I don't want to solve because it is very time consuming

I will now implement the tour locations the same way as the tour markers, when this is finished, I'll merge it into the main branch.

These are map2 related files which I've changed

![image](https://github.com/mytourbook/mytourbook/assets/1283445/9d3d7930-a5f1-4c82-ac41-6e2c907669ca)

**Reply by FJBDev** [reply]
>You should also not change related file, e.g.

>otherwise there will be a lot of merge conflicts which I don't want to solve because it is very time consuming

Ok. I will pause my development then

**Comment by FJBDev** [comment]
Also, I haven't seen this issue anymore in my 24.5 MT PROD

**Reply by wolfgang-ch** [reply]
So the above fix https://github.com/mytourbook/mytourbook/discussions/1308#discussioncomment-8948474 seems to work and the bug https://github.com/mytourbook/mytourbook/issues/1310 could be closed?

**Reply by FJBDev** [reply]
I think so, yes.
If it happens again, i will reopen the bug
